Plugins/GRPCProtobufGenerator/Plugin.swift

```diff
@@ -0,0 +1,299 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import Foundation
+import PackagePlugin
+
+// Entry-point when using Package manifest
+extension GRPCProtobufGenerator: BuildToolPlugin {
+  /// Create build commands, the entry-point when using a Package manifest.
+  func createBuildCommands(context: PluginContext, target: Target) async throws -> [Command] {
+    guard let swiftTarget = target as? SwiftSourceModuleTarget else {
+      throw PluginError.incompatibleTarget(target.name)
+    }
+    let configFiles = swiftTarget.sourceFiles(withSuffix: configFileName).map { $0.url }
+    let inputFiles = swiftTarget.sourceFiles(withSuffix: ".proto").map { $0.url }
+    return try createBuildCommands(
+      pluginWorkDirectory: context.pluginWorkDirectoryURL,
+      tool: context.tool,
+      inputFiles: inputFiles,
+      configFiles: configFiles,
+      targetName: target.name
+    )
+  }
+}
+
+#if canImport(XcodeProjectPlugin)
+import XcodeProjectPlugin
+
+// Entry-point when using Xcode projects
+extension GRPCProtobufGenerator: XcodeBuildToolPlugin {
+  /// Create build commands, the entry-point when using an Xcode project.
+  func createBuildCommands(context: XcodePluginContext, target: XcodeTarget) throws -> [Command] {
+    let configFiles = target.inputFiles.filter {
+      $0.url.lastPathComponent == configFileName
+    }.map { $0.url }
+    let inputFiles = target.inputFiles.filter { $0.url.lastPathComponent.hasSuffix(".proto") }.map {
+      $0.url
+    }
+    return try createBuildCommands(
+      pluginWorkDirectory: context.pluginWorkDirectoryURL,
+      tool: context.tool,
+      inputFiles: inputFiles,
+      configFiles: configFiles,
+      targetName: target.displayName
+    )
+  }
+}
+#endif
+
+@main
+struct GRPCProtobufGenerator {
+  /// Build plugin code common to both invocation types: package manifest Xcode project
+  func createBuildCommands(
+    pluginWorkDirectory: URL,
+    tool: (String) throws -> PluginContext.Tool,
+    inputFiles: [URL],
+    configFiles: [URL],
+    targetName: String
+  ) throws -> [Command] {
+    let configs = try readConfigFiles(configFiles, pluginWorkDirectory: pluginWorkDirectory)
+
+    let protocGenGRPCSwiftPath = try tool("protoc-gen-grpc-swift").url
+    let protocGenSwiftPath = try tool("protoc-gen-swift").url
+
+    var commands: [Command] = []
+    for inputFile in inputFiles {
+      guard let (configFilePath, config) = configs.findApplicableConfig(for: inputFile) else {
+        throw PluginError.noConfigFilesFound
+      }
+
+      let protocPath = try deriveProtocPath(using: config, tool: tool)
+      let protoDirectoryPaths: [String]
+      if !config.importPaths.isEmpty {
+        protoDirectoryPaths = config.importPaths
+      } else {
+        protoDirectoryPaths = [configFilePath.deletingLastPathComponent().absoluteStringNoScheme]
+      }
```

Comment:
nit: avoid negation if it makes sense to:

```suggestion
      if config.importPaths.isEmpty {
        protoDirectoryPaths = [configFilePath.deletingLastPathComponent().absoluteStringNoScheme]
      } else {
        protoDirectoryPaths = config.importPaths
      }
```